Twitch VOD extractor - chapters

mpeter50 opened the pull request:
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [ ] New extractor
- [x] New feature

---

### Description of your *pull request* and other information

This pull request adds 2 new capabilities to the TwitchVOD extractor:
- extracting chapters
- extracting live_chat

Chapters means game change events. The importance is that if the streamer changed to an other game, then the new game's name is seen as a new chapter from that timecode.
Live chat is the historical stream chat, it is saved as a subtitle with the language live_chat.

I also added a new test case that verifies the 2 new capabilities (at once, if that's a problem let me know), and also changed where the live chat JSON url is placed (it was already saved), but not sure if it's worth keeping because it'll be inaccessible in a few weeks.

Related discussion here: https://github.com/yt-dlp/yt-dlp/discussions/1394

mpeter50 commented:
I've left a few questions as TODOs in the code, where I was uncertain if I was doing things in the right way.

mpeter50 commented:
Also, the test case that I added is based on a stream VOD that will be inaccessible in a few weeks. Is there any special treatment that test case needs?

pukkandan commented on yt_dlp/extractor/twitch.py:
```diff
@@ -259,9 +311,19 @@ def _download_info(self, item_id):
-        video = data.get('video')
+            'Downloading stream metadata GraphQL')
+
+        # video = data[0].get('data').get('video')
```
remove

pukkandan commented on yt_dlp/extractor/twitch.py:
```diff
@@ -314,6 +376,33 @@ def _extract_info_gql(info, item_id):
+                # TODO: warning or debug log if found unknown moment type
+                continue
+
+            momentPosition = int_or_none(moment.get('positionMilliseconds'))
+            momentDuration = int_or_none(moment.get('durationMilliseconds'))
+
+            if momentPosition is None or momentDuration is None:
+                # TODO: warning or debug log about missing important data
+                chapters.clear()
+                break
+
+            momentPosition /= 1000
+            momentDuration /= 1000
```
int_or_none has a scale parameter

pukkandan commented on yt_dlp/extractor/twitch.py:
```diff
@@ -314,6 +376,33 @@ def _extract_info_gql(info, item_id):
+        chapters = list()
+        for moment in info.get('moments'):
+            moment = moment.get('node')
+            if moment.get('type') != 'GAME_CHANGE':
+                # TODO: warning or debug log if found unknown moment type
+                continue
```
why? The type doesnt affect anything in code

mpeter50 replied:
moments seem to be a multipurpose feature (but for now this might be the only type that exists), and I have a feeling that when Twitch modifies this feature, chapter extraction might silently break and I'll never notice.
probably this is silly, but I see it the way as there is a warning too when a description was not found: it hints that something might be off there.

pukkandan commented on yt_dlp/extractor/twitch.py:
```diff
@@ -314,6 +376,33 @@ def _extract_info_gql(info, item_id):
                 thumbnail = thumbnail.replace('{%s}' % p, '0')
+
+        chapters = list()
+        for moment in info.get('moments'):
```
```suggestion
        for moment in info.get('moments') or []:
```

pukkandan commented on yt_dlp/extractor/twitch.py:
```diff
@@ -259,9 +311,19 @@ def _download_info(self, item_id):
+
+        # video = data[0].get('data').get('video')
+        video = traverse_obj(data[0], ('data', 'video'))
+        video["moments"] = traverse_obj(data, (1, 'data', 'video', 'moments', 'edges'))
```
use single quotes everywhere

pukkandan commented on yt_dlp/extractor/twitch.py:
```diff
@@ -314,6 +376,33 @@ def _extract_info_gql(info, item_id):
             for p in ('width', 'height'):
                 thumbnail = thumbnail.replace('{%s}' % p, '0')
+
+        chapters = list()
```
convention is to do
```suggestion
        chapters = []
```
Also, this entire for loop looks like it can be simplified into a list comprehension

mpeter50 replied:
tried to change the loop to a list comprehension, is this how you meant?

pukkandan reviewed (changes requested):
The chapter code looks good. Can you separate out the live chat code into a new PR? Since it needs to make a lot of additional API requests, some special consideration will have to be given for how to implement it lazily

pukkandan commented:
> Also, the test case that I added is based on a stream VOD that will be inaccessible in a few weeks. Is there any special treatment that test case needs?

If possible, add a test that will not become inaccessible. If that is not possible, add `'skip': 'No longer available'` to the test. That way, the test is not actually run, but other devs reading the code can see that a similar test needs to be manually performed

mpeter50 commented:
> 
> 
> The chapter code looks good. Can you separate out the live chat code into a new PR? Since it needs to make a lot of additional API requests, some special consideration will have to be given for how to implement it lazily

Sure, new PR here: #1551

pukkandan commented on yt_dlp/extractor/twitch.py:
```diff
@@ -28,7 +28,7 @@
     update_url_query,
     url_or_none,
-    urljoin,
+    urljoin, traverse_obj, str_or_none,
```
one in a line, and sort them

pukkandan commented on yt_dlp/extractor/twitch.py:
```diff
@@ -306,6 +348,26 @@ def _extract_info(info):
     def _extract_info_gql(info, item_id):
+        def moment_to_chapter(moment):
+            moment = moment.get('node')
+            if moment.get('type') != 'GAME_CHANGE':
+                # TODO: warning or debug log if found unknown moment type
+                return None  # TODO: or empty dict?
```
> moments seem to be a multipurpose feature (but for now this might be the only type that exists), and I have a feeling that when Twitch modifies this feature, chapter extraction might silently break and I'll never notice.
probably this is silly, but I see it the way as there is a warning too when a description was not found: it hints that something might be off there.

even if they add/change a new moment type, as long as the momentPosition and duration exists, the chapter will be well-formed and it doesnt matter

pukkandan commented on yt_dlp/extractor/twitch.py:
```diff
@@ -306,6 +348,26 @@ def _extract_info(info):
+            momentDuration = int_or_none(moment.get('durationMilliseconds'), 1000)
+
+            if momentPosition is None or momentDuration is None:
+                # TODO: warning or debug log about missing important data
```
u can add a warning here

mpeter50 replied:
how should I do that?
`report_warning` from utils is not accessible, because here I'm in a static method. should I remove the static decorator, or just write to stderr?

pukkandan replied:
hm.. idk why it is static method. Let me see if there was any reason. Let it be for now

pukkandan replied:
there is no reason for it to be static. You can convert it to a normal method and add the warning

pukkandan commented on yt_dlp/extractor/twitch.py:
```diff
@@ -306,6 +348,26 @@ def _extract_info(info):
+
+            if momentPosition is None or momentDuration is None:
+                # TODO: warning or debug log about missing important data
+                chapters.clear()
```
1. `chapters` should be `None` (not `[]`) if chapter data extraction fails
2. It is bad practice to clear a nonlocal variable like this, especially in the middle of assigning it

pukkandan commented on yt_dlp/extractor/twitch.py:
```diff
@@ -259,9 +292,18 @@ def _download_info(self, item_id):
-        video = data.get('video')
+            'Downloading stream metadata GraphQL')
+
+        video = traverse_obj(data[0], ('data', 'video'))
```
```suggestion
        video = traverse_obj(data, (0, 'data', 'video'))
```

mpeter50 commented:
I have changed a few more things, too.

One such thing is that if a chapter has missing data, I add the dict to the dhapter list, that contains the original json data of that chapter. Is this ok?
The reason for this is that if the data wasn't formatted as expected, previously it would have been lost.
This way it is still saved in the infojson, so the user may be able to work with it until there's a fix.

The other thing is that somehow the check on line 393 was broken, it worked the opposite way. I fixed it in 1d145c9. No idea how the tests passed before, because the chapter entry wasn't there in the infodict since 142de849

pukkandan commented:
> One such thing is that if a chapter has missing data, I add the dict to the dhapter list, that contains the original json data of that chapter. Is this ok?

No, all data in the info_dict must adhere to the specification as explained in extractor/common.py

> The other thing is that somehow the check on line 393 was broken, it worked the opposite way. I fixed it in 1d145c9. No idea how the tests passed before, because the chapter entry wasn't there in the infodict since 142de84.

Download tests are not run in GHA due to various issues. The developers are expected to manually run them for the extractor(s) they are working on

mpeter50 commented:
> No, all data in the info_dict must adhere to the specification as explained in extractor/common.py

Ok, no problem, fixed

> Download tests are not run in GHA due to various issues. The developers are expected to manually run them for the extractor(s) they are working on

Oh, I expected that the actual downloads don't happen, but I though the infodict tests are still made. Ran the tests now, it should be ok.

mpeter50 commented:
Oh well, forgot to push.. sorry!
